File: examples/quickstart.py
"""aikstockdata quickstart — Korean stock data, no API key.

    python quickstart.py            # today's market digest
    python quickstart.py 005930     # one stock (Samsung Electronics)
    python quickstart.py 삼성전자    # look up by name, then fetch

Standard library only. No dependencies.
"""
import json
import sys
import urllib.parse
import urllib.request

BASE = "https://aikstockdata.com"


def get(path):
    """Fetch a public JSON file. A User-Agent header is required."""
    req = urllib.request.Request(BASE + path, headers={"User-Agent": "aikstockdata-example"})
    with urllib.request.urlopen(req, timeout=60) as r:
        return json.loads(r.read().decode("utf-8"))


def market_digest():
    t = get("/data/public/today.json")
    b = t.get("market_breadth") or {}
    print(f"기준일(price) {t.get('quote_as_of')} · 공시 수록 {t.get('disclosure_through')}")
    print(f"상승 {b.get('up')} · 하락 {b.get('down')} · 보합 {b.get('flat')} · 분위기 {b.get('tone')}")
    print("\n주요 공시:")
    for d in (t.get("top_disclosures") or [])[:5]:
        print(f"  [{d.get('score')}] {d.get('name')} — {d.get('label')}")
        if d.get("fact"):
            print(f"        {d['fact']}")


def find_code(name):
    """Name -> 6-digit code. The lightweight index declares URL patterns once."""
    idx = get("/data/public/search_index_min.json")
    for it in idx["items"]:
        if it["n"] == name:
            return it["c"]
    hits = [it for it in idx["items"] if name in it["n"]]
    if hits:
        return hits[0]["c"]
    raise SystemExit(f"'{name}' 를 찾지 못했습니다.")


def _won(v):
    """None stays None — never render a missing value as 0."""
    if v is None:
        return "n/a"
    return f"{v / 1e12:,.2f}조" if abs(v) >= 1e12 else f"{v / 1e8:,.0f}억"


def stock(code):
    s = get(f"/data/public/s/{code}.json")          # ~5 KB per stock
    q = s.get("quote") or {}
    print(f"{s.get('name_ko')} ({s.get('code')}) · {s.get('market')}")
    # null means 'not provided'. 0 means an actual measured zero (see has_trade).
    print(f"  종가 {q.get('close'):,}원 · 등락률 {q.get('change_pct')}% · 거래량 {q.get('volume'):,}")
    print(f"  시가총액 {_won(q.get('market_cap_krw'))} · 기준일 {q.get('as_of')}"
          f" · 거래여부 has_trade={q.get('has_trade')}")

    fin = s.get("financials") or {}
    if fin:
        print(f"  재무 {fin.get('period')} ({fin.get('basis')}) — 단위 {fin.get('unit')}")
        for key, label in (("revenue", "매출액"), ("operating_income", "영업이익"),
                           ("net_income", "순이익")):
            m = fin.get(key) or {}
            yoy = m.get("yoy_pct")
            print(f"    {label} {_won(m.get('current'))}"
                  + (f" (전년 대비 {yoy:+.1f}%)" if yoy is not None else ""))

    for d in (s.get("recent_disclosures") or [])[:3]:
        print(f"  공시 {d.get('rcept_dt')} [{d.get('type')}] {d.get('fact') or d.get('title')}")
        if d.get("dart_url"):
            print(f"        원문 {d['dart_url']}")


def check_freshness():
    """Always check before quoting numbers: status is computed, not asserted."""
    f = get("/data/public/index.json")["freshness"]
    if f["status"] != "fresh":
        print(f"⚠ 데이터 지연: {f['status']} (시세 기준일 {f['quote_as_of']}, "
              f"{f.get('quote_as_of_age_days')}일 경과)")
    return f


if __name__ == "__main__":
    check_freshness()
    arg = sys.argv[1] if len(sys.argv) > 1 else None
    if not arg:
        market_digest()
    elif arg.isdigit() and len(arg) == 6:
        stock(arg)
    else:
        stock(find_code(arg))

File: examples/quickstart.js
/**
 * aikstockdata quickstart — Korean stock data, no API key.
 *
 *   node quickstart.js            # today's market digest
 *   node quickstart.js 005930     # one stock (Samsung Electronics)
 *
 * Runs in Node 18+ and in the browser (CORS is open on /data/public/*).
 */
const BASE = "https://aikstockdata.com";

const get = async (path) => {
  const r = await fetch(BASE + path);
  if (!r.ok) throw new Error(`${path} → HTTP ${r.status}`);
  return r.json();
};

/** null means "not provided" — never render it as 0. */
const won = (v) =>
  v == null ? "n/a"
    : Math.abs(v) >= 1e12 ? `${(v / 1e12).toFixed(2)}조`
      : `${Math.round(v / 1e8).toLocaleString()}억`;

/** Freshness is computed by the publisher, not asserted. Check it before quoting numbers. */
async function checkFreshness() {
  const { freshness: f } = await get("/data/public/index.json");
  if (f.status !== "fresh") {
    console.warn(`⚠ 데이터 지연: ${f.status} (기준일 ${f.quote_as_of}, ${f.quote_as_of_age_days}일 경과)`);
  }
  return f;
}

async function marketDigest() {
  const t = await get("/data/public/today.json");
  const b = t.market_breadth || {};
  console.log(`기준일(price) ${t.quote_as_of} · 공시 수록 ${t.disclosure_through}`);
  console.log(`상승 ${b.up} · 하락 ${b.down} · 보합 ${b.flat} · 분위기 ${b.tone}\n`);
  for (const d of (t.top_disclosures || []).slice(0, 5)) {
    console.log(`  [${d.score}] ${d.name} — ${d.label}`);
    if (d.fact) console.log(`        ${d.fact}`);
  }
}

async function stock(code) {
  const s = await get(`/data/public/s/${code}.json`);   // ~5 KB per stock
  const q = s.quote || {};
  console.log(`${s.name_ko} (${s.code}) · ${s.market}`);
  console.log(`  종가 ${q.close?.toLocaleString()}원 · 등락률 ${q.change_pct}% · 거래량 ${q.volume?.toLocaleString()}`);
  console.log(`  시가총액 ${won(q.market_cap_krw)} · 기준일 ${q.as_of} · has_trade=${q.has_trade}`);

  const f = s.financials;
  if (f) {
    console.log(`  재무 ${f.period} (${f.basis})`);
    for (const [k, label] of [["revenue", "매출액"], ["operating_income", "영업이익"], ["net_income", "순이익"]]) {
      const m = f[k] || {};
      const yoy = m.yoy_pct == null ? "" : ` (전년 대비 ${m.yoy_pct > 0 ? "+" : ""}${m.yoy_pct}%)`;
      console.log(`    ${label} ${won(m.current)}${yoy}`);
    }
  }
  for (const d of (s.recent_disclosures || []).slice(0, 3)) {
    console.log(`  공시 ${d.rcept_dt} [${d.type}] ${d.fact || d.title}`);
  }
}

/** Name -> code. The lightweight index declares URL patterns once instead of repeating them. */
async function findCode(name) {
  const idx = await get("/data/public/search_index_min.json");
  const hit = idx.items.find((i) => i.n === name) || idx.items.find((i) => i.n.includes(name));
  if (!hit) throw new Error(`'${name}' 를 찾지 못했습니다.`);
  return hit.c;
}

(async () => {
  await checkFreshness();
  const arg = process.argv[2];
  if (!arg) return marketDigest();
  return stock(/^\d{6}$/.test(arg) ? arg : await findCode(arg));
})().catch((e) => {
  console.error(e.message);
  process.exit(1);
});
